pkg/fake_gcloud/lib/retry_enforcer_storage.dart

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:gcloud/storage.dart';
+
+void _verifyRetryOnStack() {
+  final st = StackTrace.current.toString();
+  if (st.contains('package:retry/')) return;
+  if (st.contains('retryAsync')) return; // lib/shared/utils.dart
+
+  // detect direct test calls
+  final linesWithoutThisFile = st
+      .split('\n')
+      .where((l) => !l.contains('retry_enforcer_storage.dart'))
+      .toList();
+  if (linesWithoutThisFile.isNotEmpty &&
+      linesWithoutThisFile.first.contains('_test.dart')) {
+    return;
+  }
+
+  print('Missing retry detected:\n$st\n');
+  throw AssertionError('retry is not present in stacktrace: $st');
```

Comment:
Should we also detect that we only have a single retry call on the stack?

Reply:
I think we should allow multiple re-entry in the retry stack, e.g. one for a high-level iteration of items, while the second one may be narrower operations on it.

We could ignore re-entry detection in the retry stack if we could check if the current call is sufficiently close to the latest retry invocation. But I would consider that check to come later, after the low-hanging fixes (e.g. no retry at all) has been done.

Reply:
> I think we should allow multiple re-entry in the retry stack, e.g. one for a high-level iteration of items, while the second one may be narrower operations on it.

Yeah - we would need to have some kind of design here. But nested retries easily leads to n^2 attempts...

> But I would consider that check to come later, 

SG.